File: extract_cbt.py
import sys
import json
import os
import re
import base64

try:
    import fitz # PyMuPDF
except ImportError:
    print(json.dumps({"error": "PyMuPDF not installed"}))
    sys.exit(1)

def is_hindi(text):
    # Check if text contains Devanagari characters
    return bool(re.search(r'[\u0900-\u097F]', text))

def parse_pdf(pdf_path):
    if not os.path.exists(pdf_path):
        print(f"Error: File not found: {pdf_path}")
        sys.exit(1)
        
    doc = fitz.open(pdf_path)
    
    questions = []
    metadata = {
        "exam_name": "Mock CBT Exam",
        "total_questions": 0,
        "total_marks": 0,
        "duration_minutes": 180,
        "subjects": ["Physics", "Chemistry", "Mathematics"]
    }
    
    validation_report = {
        "total_questions": 0,
        "missing_answers": 0,
        "low_confidence_flags": 0,
        "unresolved_images": 0
    }
    
    current_subject = "Physics"
    figures_list = []
    
    # Store all english text blocks with their coordinates
    full_text_blocks = []
    
    for i in range(len(doc)):
        page = doc[i]
        
        # Extract images
        image_list = page.get_images(full=True)
        for img_index, img in enumerate(image_list):
            xref = img[0]
            base_image = doc.extract_image(xref)
            image_bytes = base_image["image"]
            ext = base_image["ext"]
            rects = page.get_image_rects(xref)
            if rects:
                rect = rects[0]
                b64 = base64.b64encode(image_bytes).decode('utf-8')
                figures_list.append({
                    "page": i,
                    "y0": rect.y0,
                    "y1": rect.y1,
                    "b64": f"data:image/{ext};base64,{b64}",
                    "used": False
                })
        
        # Extract text
        blocks = page.get_text("dict")["blocks"]
        for b in blocks:
            if b["type"] != 0: continue
            
            block_text = ""
            bbox = b["bbox"]
            for line in b["lines"]:
                line_text = ""
                has_hindi = False
                for span in line["spans"]:
                    text = span["text"]
                    if is_hindi(text):
                        has_hindi = True
                        break
                    line_text += text
                
                if not has_hindi and line_text.strip():
                    block_text += line_text + "\n"
            
            if block_text.strip():
                full_text_blocks.append({
                    "page": i,
                    "bbox": bbox,
                    "text": block_text.strip()
                })

    combined_text = "\n".join([b["text"] for b in full_text_blocks])
    
    q_pattern = re.compile(r'^(\d+)\.\s+(.*)', re.MULTILINE)
    
    matches = list(q_pattern.finditer(combined_text))
        
    for idx, match in enumerate(matches):
        q_num = match.group(1)
        start_pos = match.start()
        end_pos = matches[idx+1].start() if idx + 1 < len(matches) else len(combined_text)
        
        q_chunk = combined_text[start_pos:end_pos]
        
        if "CHEMISTRY" in q_chunk.upper(): current_subject = "Chemistry"
        elif "MATHEMATICS" in q_chunk.upper(): current_subject = "Mathematics"
        
        opt_a = re.search(r'\(A\)\s*(.*?)(?=\(B\)|\(C\)|\(D\)|\n\n|$)', q_chunk, re.DOTALL)
        opt_b = re.search(r'\(B\)\s*(.*?)(?=\(C\)|\(D\)|\n\n|$)', q_chunk, re.DOTALL)
        opt_c = re.search(r'\(C\)\s*(.*?)(?=\(D\)|\n\n|$)', q_chunk, re.DOTALL)
        opt_d = re.search(r'\(D\)\s*(.*?)(?=\n\n|$)', q_chunk, re.DOTALL)
        
        options = []
        q_type = "numerical"
        if opt_a and opt_b and opt_c and opt_d:
            q_type = "single_correct"
            options = [
                {"label": "A", "text": opt_a.group(1).strip(), "image": None},
                {"label": "B", "text": opt_b.group(1).strip(), "image": None},
                {"label": "C", "text": opt_c.group(1).strip(), "image": None},
                {"label": "D", "text": opt_d.group(1).strip(), "image": None}
            ]
        
        if q_type == "single_correct":
            q_text_end = q_chunk.find("(A)")
            q_text = q_chunk[len(q_num)+1:q_text_end].strip()
        else:
            q_text = q_chunk[len(q_num)+1:].strip()
            
        q_id = f"q-{q_num}-{idx}"
        
        q_images = []
        for fig in figures_list:
            if not fig["used"]:
                q_images.append(fig["b64"])
                fig["used"] = True
                break
                
        questions.append({
            "question_id": q_id,
            "subject": current_subject,
            "section": "Section-I",
            "question_number": int(q_num),
            "question_text": q_text,
            "question_images": q_images,
            "options": options,
            "question_type": q_type,
            "marks_correct": 4,
            "marks_incorrect": -1 if q_type == "single_correct" else 0,
            "correct_answer": "A",
            "solution": None
        })

    metadata["total_questions"] = len(questions)
    metadata["total_marks"] = len(questions) * 4
    
    validation_report["total_questions"] = len(questions)
    validation_report["missing_answers"] = len([q for q in questions if not q["correct_answer"]])
    
    output = {
        "metadata": metadata,
        "questions": questions
    }
    
    with open("test_data.json", "w", encoding="utf-8") as f:
        json.dump(output, f, indent=2, ensure_ascii=False)
        
    with open("validation_report.md", "w", encoding="utf-8") as f:
        f.write("# Validation Report\n\n")
        f.write(f"- Total Questions Found: {validation_report['total_questions']}\n")
        f.write(f"- Questions with Missing Answers: {validation_report['missing_answers']}\n")
        f.write(f"- Questions Flagged Low-Confidence: {validation_report['low_confidence_flags']}\n")
        f.write(f"- Questions with Unresolved Images: {validation_report['unresolved_images']}\n")

    print(f"Extraction complete! Found {len(questions)} questions.")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python extract_cbt.py <pdf_path>")
        sys.exit(1)
    parse_pdf(sys.argv[1])

File: extract_cbt.js
const fs = require('fs');
const path = require('path');
const pdfjsLib = require('pdfjs-dist/legacy/build/pdf.js');

const isHindi = (text) => /[\u0900-\u097F]/.test(text);

async function parsePdf() {
  const filePath = 'C:/Users/rahul/OneDrive/Documents/Question_Report_130.pdf';
  if (!fs.existsSync(filePath)) {
    console.error(`Error: ${filePath} not found.`);
    process.exit(1);
  }
  
  const buffer = fs.readFileSync(filePath);
  const data = new Uint8Array(buffer);
  
  const standardFontDataUrl = path.join(__dirname, 'node_modules/pdfjs-dist/standard_fonts/');
  const loadingTask = pdfjsLib.getDocument({ 
    data, 
    verbosity: 0,
    standardFontDataUrl,
  });
  
  const pdfDocument = await loadingTask.promise;
  
  let fullText = "";
  
  for (let i = 1; i <= pdfDocument.numPages; i++) {
    const page = await pdfDocument.getPage(i);
    const textContent = await page.getTextContent();
    const viewport = page.getViewport({ scale: 1 });
    const items = textContent.items;
    
    items.sort((a, b) => {
        const yA = a.transform[5];
        const yB = b.transform[5];
        if (Math.abs(yA - yB) < 4) return a.transform[4] - b.transform[4];
        return yB - yA; // Flipped Y
    });
    
    let lastY = -1;
    for (const item of items) {
        if (isHindi(item.str)) {
           continue;
        }
        
        // Skip right column if it exists (usually x > width/2)
        // Let's assume standard A4 width is ~595. If X > 300, it might be right column
        const x = item.transform[4];
        if (x > viewport.width * 0.52) {
           continue; 
        }

        if (lastY !== -1 && Math.abs(item.transform[5] - lastY) > 8) {
            fullText += "\n";
        } else if (lastY !== -1) {
            fullText += " ";
        }
        fullText += item.str.trim();
        lastY = item.transform[5];
    }
    fullText += "\n\n";
  }

  fullText = fullText.replace(/ {2,}/g, ' ');

  // Clean up headers and footers
  fullText = fullText.replace(/LEADER TEST SERIES \/ JOINT PACKAGE COURSE[\s\S]*?@JEEAdvanced_2025/gi, '');
  fullText = fullText.replace(/Space for Rough Work/gi, '');

  // Output intermediate for debugging
  fs.writeFileSync('debug_fulltext.txt', fullText, 'utf-8');

  const questions = [];
  const metadata = {
      exam_name: "Mock CBT Exam",
      total_questions: 0,
      total_marks: 0,
      duration_minutes: 180,
      subjects: ["Physics", "Chemistry", "Mathematics"]
  };
  
  const validation_report = {
      total_questions: 0,
      missing_answers: 0,
      low_confidence_flags: 0,
      unresolved_images: 0
  };

  const chunks = fullText.split(/\(A\)/);
  let currentSubject = "Physics";

  for (let idx = 1; idx < chunks.length; idx++) {
      const prevChunk = chunks[idx - 1];
      const currentChunk = chunks[idx];
      
      let qText = prevChunk;
      const lastD = qText.lastIndexOf('(D)');
      if (lastD !== -1) {
          qText = qText.substring(lastD + 3);
      }
      
      if (qText.toUpperCase().includes("CHEMISTRY")) currentSubject = "Chemistry";
      if (qText.toUpperCase().includes("MATHEMATICS")) currentSubject = "Mathematics";

      qText = qText.trim();
      
      let qNumStr = "0";
      const qNumMatch = qText.match(/^(\d+)[\.\)]/);
      if (qNumMatch) {
          qNumStr = qNumMatch[1];
          qText = qText.replace(/^[\d\.\)]+\s*/, '').trim();
      } else {
          qNumStr = String(idx);
      }

      const bSplit = currentChunk.split(/\(B\)/);
      if (bSplit.length < 2) continue;
      
      const optionA = bSplit[0].trim();
      const cSplit = bSplit[1].split(/\(C\)/);
      if (cSplit.length < 2) continue;
      
      const optionB = cSplit[0].trim();
      const dSplit = cSplit[1].split(/\(D\)/);
      if (dSplit.length < 2) continue;
      
      const optionC = dSplit[0].trim();
      const optionD = dSplit[1].split('\n')[0].trim() || dSplit[1].substring(0, 50).trim();
      
      if (qText.length > 3) {
          questions.push({
              question_id: `q-${idx}`,
              subject: currentSubject,
              section: "Section-I",
              question_number: parseInt(qNumStr) || idx,
              question_text: qText,
              question_images: [],
              options: [
                  { label: "A", text: optionA, image: null },
                  { label: "B", text: optionB, image: null },
                  { label: "C", text: optionC, image: null },
                  { label: "D", text: optionD, image: null }
              ],
              question_type: "single_correct",
              marks_correct: 4,
              marks_incorrect: -1,
              correct_answer: "A",
              solution: null
          });
      }
  }

  metadata.total_questions = questions.length;
  metadata.total_marks = questions.length * 4;
  validation_report.total_questions = questions.length;
  validation_report.missing_answers = questions.filter(q => !q.correct_answer).length;

  const output = { metadata, questions };
  
  fs.writeFileSync('app/(cbt)/tests/cbt-demo/test_data.json', JSON.stringify(output, null, 2), 'utf-8');
  fs.writeFileSync('validation_report.md', 'Generated...', 'utf-8');
  
  console.log(`Extraction complete! Found ${questions.length} questions.`);
}

parsePdf().catch(console.error);
